<a href="https://colab.research.google.com/github/Letch49/ML_vvsu_2024/blob/master/lab1/%D0%97%D0%B0%D0%B4%D0%B0%D0%BD%D0%B8%D0%B5_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Задание по практике 1

##Загрузка датасета
В файле classification datasets расположены ссылки на датасеты. Выберите один в соответствии со своим вариантом, скачайте и прочитайте файл

**Не забудьте загрузить нужные библиотеки**

Например: pip3 install pandas

In [1]:
import pandas as pd

In [4]:
df = pd.read_csv('/content/global-data-on-sustainable-energy.csv')
df

Entity  Year  Access to electricity (% of population)  \
0     Afghanistan  2000                                 1.613591   
1     Afghanistan  2001                                 4.074574   
2     Afghanistan  2002                                 9.409158   
3     Afghanistan  2003                                14.738506   
4     Afghanistan  2004                                20.064968   
...           ...   ...                                      ...   
3644     Zimbabwe  2016                                42.561730   
3645     Zimbabwe  2017                                44.178635   
3646     Zimbabwe  2018                                45.572647   
3647     Zimbabwe  2019                                46.781475   
3648     Zimbabwe  2020                                52.747670   

      Access to clean fuels for cooking  \
0                                   6.2   
1                                   7.2   
2                                   8.2   
3                                   9.5   
4                                  10.9   
...                                 ...   
3644                               29.8   
3645                               29.8   
3646                               29.9   
3647                               30.1   
3648                               30.4   

      Renewable-electricity-generating-capacity-per-capita  \
0                                                  9.22      
1                                                  8.86      
2                                                  8.47      
3                                                  8.09      
4                                                  7.75      
...                                                 ...      
3644                                              62.88      
3645                                              62.33      
3646                                              82.53      
3647                                              81.40      
3648                                              80.61      

      Financial flows to developing countries (US $)  \
0                                            20000.0   
1                                           130000.0   
2                                          3950000.0   
3                                         25970000.0   
4                                                NaN   
...                                              ...   
3644                                         30000.0   
3645                                       5570000.0   
3646                                         10000.0   
3647                                        250000.0   
3648                                         30000.0   

      Renewable energy share in the total final energy consumption (%)  \
0                                                 44.99                  
1                                                 45.60                  
2                                                 37.83                  
3                                                 36.66                  
4                                                 44.24                  
...                                                 ...                  
3644                                              81.90                  
3645                                              82.46                  
3646                                              80.23                  
3647                                              81.50                  
3648                                              81.90                  

      Electricity from fossil fuels (TWh)  Electricity from nuclear (TWh)  \
0                                    0.16                             0.0   
1                                    0.09                             0.0   
2                                    0.13                             0.0   
3                                    0.31                     

##Первое знакомство с датасетом
В следующих ячейках выведите:


1.   Колонки датасета
2.   Первые 5 значений
3.   Последние 5 значений
4.   Статистику по числовым колонкам



In [5]:
df.columns

Index(['Entity', 'Year', 'Access to electricity (% of population)',
       'Access to clean fuels for cooking',
       'Renewable-electricity-generating-capacity-per-capita',
       'Financial flows to developing countries (US $)',
       'Renewable energy share in the total final energy consumption (%)',
       'Electricity from fossil fuels (TWh)', 'Electricity from nuclear (TWh)',
       'Electricity from renewables (TWh)',
       'Low-carbon electricity (% electricity)',
       'Primary energy consumption per capita (kWh/person)',
       'Energy intensity level of primary energy (MJ/$2017 PPP GDP)',
       'Value_co2_emissions_kt_by_country',
       'Renewables (% equivalent primary energy)', 'gdp_growth',
       'gdp_per_capita', 'Density\n(P/Km2)', 'Land Area(Km2)', 'Latitude',
       'Longitude'],
      dtype='object')

In [6]:
df.head(5)

Entity  Year  Access to electricity (% of population)  \
0  Afghanistan  2000                                 1.613591   
1  Afghanistan  2001                                 4.074574   
2  Afghanistan  2002                                 9.409158   
3  Afghanistan  2003                                14.738506   
4  Afghanistan  2004                                20.064968   

   Access to clean fuels for cooking  \
0                                6.2   
1                                7.2   
2                                8.2   
3                                9.5   
4                               10.9   

   Renewable-electricity-generating-capacity-per-capita  \
0                                               9.22      
1                                               8.86      
2                                               8.47      
3                                               8.09      
4                                               7.75      

   Financial flows to developing countries (US $)  \
0                                         20000.0   
1                                        130000.0   
2                                       3950000.0   
3                                      25970000.0   
4                                             NaN   

   Renewable energy share in the total final energy consumption (%)  \
0                                              44.99                  
1                                              45.60                  
2                                              37.83                  
3                                              36.66                  
4                                              44.24                  

   Electricity from fossil fuels (TWh)  Electricity from nuclear (TWh)  \
0                                 0.16                             0.0   
1                                 0.09                             0.0   
2                                 0.13                             0.0   
3                                 0.31                             0.0   
4                                 0.33                             0.0   

   Electricity from renewables (TWh)  ...  \
0                               0.31  ...   
1                               0.50  ...   
2                               0.56  ...   
3                               0.63  ...   
4                               0.56  ...   

   Primary energy consumption per capita (kWh/person)  \
0                                          302.59482    
1                                          236.89185    
2                                          210.86215    
3                                          229.96822    
4                                          204.23125    

   Energy intensity level of primary energy (MJ/$2017 PPP GDP)  \
0                                               1.64             
1                                               1.74             
2                                               1.40             
3                                               1.40             
4                                               1.20             

   Value_co2_emissions_kt_by_country  \
0                         760.000000   
1                         730.000000   
2                        1029.999971   
3                        1220.000029   
4                        1029.999971   

   Renewables (% equivalent primary energy)  gdp_growth  gdp_per_capita  \
0                                       NaN         NaN             NaN   
1                                       NaN         NaN             NaN   
2                                       NaN         NaN      179.426579   
3                                       NaN    8.832278      190.683814   
4                                       NaN    1.414118      211.382074   

   Density\n(P/Km2) Land Area(Km2)  Latitude  Longitude  
0                60       652230.0  33.93911  67.709953  
1          

In [7]:
df[-5:]

Entity  Year  Access to electricity (% of population)  \
3644  Zimbabwe  2016                                42.561730   
3645  Zimbabwe  2017                                44.178635   
3646  Zimbabwe  2018                                45.572647   
3647  Zimbabwe  2019                                46.781475   
3648  Zimbabwe  2020                                52.747670   

      Access to clean fuels for cooking  \
3644                               29.8   
3645                               29.8   
3646                               29.9   
3647                               30.1   
3648                               30.4   

      Renewable-electricity-generating-capacity-per-capita  \
3644                                              62.88      
3645                                              62.33      
3646                                              82.53      
3647                                              81.40      
3648                                              80.61      

      Financial flows to developing countries (US $)  \
3644                                         30000.0   
3645                                       5570000.0   
3646                                         10000.0   
3647                                        250000.0   
3648                                         30000.0   

      Renewable energy share in the total final energy consumption (%)  \
3644                                              81.90                  
3645                                              82.46                  
3646                                              80.23                  
3647                                              81.50                  
3648                                              81.90                  

      Electricity from fossil fuels (TWh)  Electricity from nuclear (TWh)  \
3644                                 3.50                             0.0   
3645                                 3.05                             0.0   
3646                                 3.73                             0.0   
3647                                 3.66                             0.0   
3648                                 3.40                             0.0   

      Electricity from renewables (TWh)  ...  \
3644                               3.32  ...   
3645                               4.30  ...   
3646                               5.46  ...   
3647                               4.58  ...   
3648                               4.19  ...   

      Primary energy consumption per capita (kWh/person)  \
3644                                          3227.6802    
3645                                          3068.0115    
3646                                          3441.9858    
3647                                          3003.6553    
3648                                          2680.1318    

      Energy intensity level of primary energy (MJ/$2017 PPP GDP)  \
3644                                              10.00             
3645                                               9.51             
3646                                               9.83             
3647                                              10.47             
3648                                              10.00             

      Value_co2_emissions_kt_by_country  \
3644                        11020.00046   
3645                        10340.00015   
3646                        12380.00011   
3647                        11760.00023   
3648                                NaN   

      Renewables (% equivalent primary energy)  gdp_growth  gdp_per_capita  \
3644                                       NaN    0.755869     1464.588957   
3645                                       NaN    4.709492     1235.189032   
3646                                       NaN    4.824211     1254.642265   
3647                                       NaN   -6.144236     1316.740657   
3648                      

In [8]:
df.describe()

Year  Access to electricity (% of population)  \
count  3649.000000                              3639.000000   
mean   2010.038367                                78.933702   
std       6.054228                                30.275541   
min    2000.000000                                 1.252269   
25%    2005.000000                                59.800890   
50%    2010.000000                                98.361570   
75%    2015.000000                               100.000000   
max    2020.000000                               100.000000   

       Access to clean fuels for cooking  \
count                        3480.000000   
mean                           63.255287   
std                            39.043658   
min                             0.000000   
25%                            23.175000   
50%                            83.150000   
75%                           100.000000   
max                           100.000000   

       Renewable-electricity-generating-capacity-per-capita  \
count                                        2718.000000      
mean                                          113.137498      
std                                           244.167256      
min                                             0.000000      
25%                                             3.540000      
50%                                            32.910000      
75%                                           112.210000      
max                                          3060.190000      

       Financial flows to developing countries (US $)  \
count                                    1.560000e+03   
mean                                     9.422400e+07   
std                                      2.981544e+08   
min                                      0.000000e+00   
25%                                      2.600000e+05   
50%                                      5.665000e+06   
75%                                      5.534750e+07   
max                                      5.202310e+09   

       Renewable energy share in the total final energy consumption (%)  \
count                                        3455.000000                  
mean                                           32.638165                  
std                                            29.894901                  
min                                             0.000000                  
25%                                             6.515000                  
50%                                            23.300000                  
75%                                            55.245000                  
max                                            96.040000                  

       Electricity from fossil fuels (TWh)  Electricity from nuclear (TWh)  \
count                          3628.000000                     3523.000000   
mean                             70.365003                       13.450190   
std                             348.051866                       73.006623   
min                               0.000000                        0.000000   
25%                               0.290000                        0.000000   
50%                               2.970000                        0.000000   
75%                              26.837500                        0.000000   
max                            5184.130000                      809.410000   

       Electricity from renewables (TWh)  \
count                        3628.000000   
mean                           23.968010   
std                           104.431085   
min                             0.000000   
25%                             0.040000   
50%                             1.470000   
75%                             9.600000   
max                          2184.940000   

       Low-carbon electricity (% electricity)  \
count                             3607.000000   
mean                                36.801182   
std                                 34.31488

**Ответьте на следующие вопросы:**


1.   Какие средние значения в каждой числовой колонке вашего датасета?
2.   Минимальные и максимальные значения в каждой числовой колонке вашего датасета





1.   **Ответ 1.** 2010.038367	78.933702	63.255287	113.137498	9.422400e+07	32.638165	70.365003	13.450190	23.968010	36.801182	25743.981745	5.307345	1.598665e+05	11.986707	3.441610	13283.774348	6.332135e+05	18.246388	14.822695
2.   **Ответ 2.** **min** 2000.000000	1.252269	0.000000	0.000000	0.000000e+00	0.000000	0.000000	0.000000	0.000000	0.000000	0.000000	0.110000	1.000000e+01	0.000000	-62.075920	111.927225	2.100000e+01	-40.900557	-175.198242

**max** 2020.000000	100.000000	100.000000	3060.190000	5.202310e+09	96.040000	5184.130000	809.410000	2184.940000	100.000010	262585.700000	32.570000	1.070722e+07	86.836586	123.139555	123514.196700	9.984670e+06	64.963051	178.065032

##Типизация
Присвойте соответствующим колонкам один из типов (category, float, int)

In [11]:
df['Entity'] = df['Entity'].astype('category')
df['Year'] = df['Year'].astype(int)
df['Renewable-electricity-generating-capacity-per-capita'] = df['Renewable-electricity-generating-capacity-per-capita'].astype(float)
#Вывести типы колонок тут
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3649 entries, 0 to 3648
Data columns (total 21 columns):
 #   Column                                                            Non-Null Count  Dtype   
---  ------                                                            --------------  -----   
 0   Entity                                                            3649 non-null   category
 1   Year                                                              3649 non-null   int64   
 2   Access to electricity (% of population)                           3639 non-null   float64 
 3   Access to clean fuels for cooking                                 3480 non-null   float64 
 4   Renewable-electricity-generating-capacity-per-capita              2718 non-null   float64 
 5   Financial flows to developing countries (US $)                    1560 non-null   float64 
 6   Renewable energy share in the total final energy consumption (%)  3455 non-null   float64 
 7   Electricity from fossil 

Какие признаки вы можете выделить в своем датасете (категориальные, числовые и др., **минимум 3 разных**, помните что можно также выделить признаки более точно, например числовые - порядковые, количественные и т.д.)?
Ответ. Entity - категориальный, year - кол-ные, Latitude - кол-ные


##Фильтрация
Офильтруйте 3 колонки вашего датасета и выведите результат

In [13]:
df.loc[df['Latitude'] > 24]

Entity  Year  Access to electricity (% of population)  \
0     Afghanistan  2000                                 1.613591   
1     Afghanistan  2001                                 4.074574   
2     Afghanistan  2002                                 9.409158   
3     Afghanistan  2003                                14.738506   
4     Afghanistan  2004                                20.064968   
...           ...   ...                                      ...   
3560   Uzbekistan  2016                               100.000000   
3561   Uzbekistan  2017                               100.000000   
3562   Uzbekistan  2018                               100.000000   
3563   Uzbekistan  2019                               100.000000   
3564   Uzbekistan  2020                               100.000000   

      Access to clean fuels for cooking  \
0                                  6.20   
1                                  7.20   
2                                  8.20   
3                                  9.50   
4                                 10.90   
...                                 ...   
3560                              85.20   
3561                              84.90   
3562                              84.30   
3563                              84.60   
3564                              84.25   

      Renewable-electricity-generating-capacity-per-capita  \
0                                                  9.22      
1                                                  8.86      
2                                                  8.47      
3                                                  8.09      
4                                                  7.75      
...                                                 ...      
3560                                              59.88      
3561                                              58.24      
3562                                              59.09      
3563                                              57.96      
3564                                              60.02      

      Financial flows to developing countries (US $)  \
0                                            20000.0   
1                                           130000.0   
2                                          3950000.0   
3                                         25970000.0   
4                                                NaN   
...                                              ...   
3560                                        690000.0   
3561                                      60130000.0   
3562                                      84130000.0   
3563                                      65940000.0   
3564                                             NaN   

      Renewable energy share in the total final energy consumption (%)  \
0                                                 44.99                  
1                                                 45.60                  
2                                                 37.83                  
3                                                 36.66                  
4                                                 44.24                  
...                                                 ...                  
3560                                               1.61                  
3561                                               1.75                  
3562                                               1.49                  
3563                                               1.57                  
3564                                                NaN                  

      Electricity from fossil fuels (TWh)  Electricity from nuclear (TWh)  \
0                                    0.16                             0.0   
1                                    0.09                             0.0   
2                                    0.13                             0.0   
3                                    0.31                     

In [17]:
df.loc[df['Year'] >= 2019]

Entity  Year  Access to electricity (% of population)  \
19    Afghanistan  2019                                97.700000   
20    Afghanistan  2020                                97.700000   
40        Albania  2019                               100.000000   
41        Albania  2020                               100.000000   
61        Algeria  2019                                99.500000   
...           ...   ...                                      ...   
3606        Yemen  2020                                73.757930   
3626       Zambia  2019                                43.000000   
3627       Zambia  2020                                44.524475   
3647     Zimbabwe  2019                                46.781475   
3648     Zimbabwe  2020                                52.747670   

      Access to clean fuels for cooking  \
19                                 31.9   
20                                 33.2   
40                                 80.7   
41                                 81.3   
61                                 99.7   
...                                 ...   
3606                               61.5   
3626                               11.2   
3627                               10.4   
3647                               30.1   
3648                               30.4   

      Renewable-electricity-generating-capacity-per-capita  \
19                                                 9.58      
20                                                 9.35      
40                                                  NaN      
41                                                  NaN      
61                                                15.93      
...                                                 ...      
3606                                               8.48      
3626                                             142.11      
3627                                             138.23      
3647                                              81.40      
3648                                              80.61      

      Financial flows to developing countries (US $)  \
19                                         4620000.0   
20                                               NaN   
40                                               NaN   
41                                               NaN   
61                                           60000.0   
...                                              ...   
3606                                             NaN   
3626                                      78570000.0   
3627                                             NaN   
3647                                        250000.0   
3648                                         30000.0   

      Renewable energy share in the total final energy consumption (%)  \
19                                                18.51                  
20                                                  NaN                  
40                                                  NaN                  
41                                                  NaN                  
61                                                 0.16                  
...                                                 ...                  
3606                                                NaN                  
3626                                              84.50                  
3627                                                NaN                  
3647                                              81.50                  
3648                                              81.90                  

      Electricity from fossil fuels (TWh)  Electricity from nuclear (TWh)  \
19                                   0.18                             0.0   
20                                   0.12                             0.0   
40                                   0.00                             0.0   
41                                   0.00                     

In [18]:
df.loc[df['Entity'] == 'Yemen']

Entity  Year  Access to electricity (% of population)  \
3586  Yemen  2000                                49.187256   
3587  Yemen  2001                                50.341240   
3588  Yemen  2002                                51.487637   
3589  Yemen  2003                                52.628803   
3590  Yemen  2004                                49.590000   
3591  Yemen  2005                                54.904820   
3592  Yemen  2006                                55.800335   
3593  Yemen  2007                                57.188570   
3594  Yemen  2008                                58.339930   
3595  Yemen  2009                                59.501240   
3596  Yemen  2010                                60.675194   
3597  Yemen  2011                                61.864506   
3598  Yemen  2012                                52.410000   
3599  Yemen  2013                                75.600000   
3600  Yemen  2014                                66.100000   
3601  Yemen  2015                                67.310930   
3602  Yemen  2016                                68.859505   
3603  Yemen  2017                                79.200000   
3604  Yemen  2018                                62.000000   
3605  Yemen  2019                                72.751076   
3606  Yemen  2020                                73.757930   

      Access to clean fuels for cooking  \
3586                              55.70   
3587                              56.10   
3588                              56.70   
3589                              57.15   
3590                              57.40   
3591                              57.90   
3592                              58.20   
3593                              58.70   
3594                              59.00   
3595                              59.50   
3596                              59.70   
3597                              60.00   
3598                              60.00   
3599                              60.30   
3600                              60.80   
3601                              61.10   
3602                              61.20   
3603                              61.35   
3604                              61.50   
3605                              61.30   
3606                              61.50   

      Renewable-electricity-generating-capacity-per-capita  \
3586                                               0.00      
3587                                               0.00      
3588                                               0.00      
3589                                               0.00      
3590                                               0.00      
3591                                               0.00      
3592                                               0.00      
3593                                               0.00      
3594                                               0.04      
3595                                               0.04      
3596                                               0.04      
3597                                               0.05      
3598                                               0.06      
3599                                               0.09      
3600                                               0.19      
3601                                               2.26      
3602                                               2.94      
3603                                               3.59      
3604                                               8.77      
3605                                               8.57      
3606                                               8.48      

      Financial flows to developing countries (US $)  \
3586                                       1480000.0   
3587                                         30000.0   
3588                                             0.0   
3589                                             NaN   
3590                                        1100

##Группировка
Сгруппируйте датасет по какому-либо категориальному признаку и выведите среднее значение численных колонок (см. метод mean в pandas)

In [20]:
df.groupby('Entity').mean()

<ipython-input-20-8223fa95397a>:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  df.groupby('Entity').mean()


Year  Access to electricity (% of population)  \
Entity                                                                 
Afghanistan          2010.0                                52.520488   
Albania              2010.0                                99.981429   
Algeria              2010.0                                99.108146   
Angola               2010.0                                35.138564   
Antigua and Barbuda  2010.0                                98.823761   
...                     ...                                      ...   
Uzbekistan           2010.0                                99.848840   
Vanuatu              2010.0                                41.269784   
Yemen                2010.0                                60.928522   
Zambia               2010.0                                27.753585   
Zimbabwe             2010.0                                39.172120   

                     Access to clean fuels for cooking  \
Entity                                                   
Afghanistan                                  19.802381   
Albania                                      63.592857   
Algeria                                      98.904762   
Angola                                       44.933333   
Antigua and Barbuda                         100.000000   
...                                                ...   
Uzbekistan                                   84.976190   
Vanuatu                                      12.050000   
Yemen                                        59.290476   
Zambia                                       14.471429   
Zimbabwe                                     30.930952   

                     Renewable-electricity-generating-capacity-per-capita  \
Entity                                                                      
Afghanistan                                                   8.581429      
Albania                                                            NaN      
Algeria                                                       9.390000      
Angola                                                       43.951429      
Antigua and Barbuda                                          22.455238      
...                                                                ...      
Uzbekistan                                                   61.972381      
Vanuatu                                                      19.339524      
Yemen                                                         1.672381      
Zambia                                                      146.577619      
Zimbabwe                                                     66.068571      

                     Financial flows to developing countries (US $)  \
Entity                                                                
Afghanistan                                            3.296947e+07   
Albania                                                         NaN   
Algeria                                                5.056250e+05   
Angola                                                 9.525000e+06   
Antigua and Barbuda                                    9.700000e+06   
...                                                             ...   
Uzbekistan                                             5.868556e+07   
Vanuatu                                                5.694000e+06   
Yemen                                                  1.059562e+07   
Zambia                                                 1.881594e+08   
Zimbabwe                                               2.651154e+07   

                     Renewable energy share in the total final energy consumption (%)  \
Entity                                                                                  
Afghanistan                                                    25.6805                  
Albania                                                            NaN                  
Algeria                                          

##Задание для гениев *
Агрегируйте числовой признак по категориальному признаку с использованием метода **без использования** библиотеки pandas, и вычислите медиану и среднее значение для каждой категории.

In [ ]:
#Агрегация